datasource/src/main/scala/quasar/physical/s3/package.scala

```diff
@@ -45,4 +53,25 @@ object S3JsonParsing {
 
 package object s3 {
   val datasourceKind: DatasourceType = DatasourceType("s3", 1L)
+
+  @SuppressWarnings(Array("org.wartremover.warts.Recursion"))
+  def resourceToDisposable[F[_]: Monad: MonadError_[?[_], Throwable], A](r: Resource[F, A])
+      : F[Disposable[F, A]] =
+    r match {
+      case Resource.Allocate(a) => a map {
+        case (res, release) => Disposable(res, release(ExitCase.Completed))
+      }
+      case Resource.Bind(src, fs) => {
```

Comment:
I'd appreciate a closer review on this piece of code. It was tricky to get it right.